### ISSUE TITLE
Recover from renderer / browser crashes

### ISSUE THREAD

**Issue by brian-mann** [cypress-io/cypress#349]
Related to #348.

It is actually possible for Cypress to implement strategies when the renderer (or browser process) crashes during a test run - something like `recoverFromRendererCrashes: true` by default.

There is already a mechanism for Cypress to "reload" mid-run, rebuild the state of every previous run test, skip over previously run tests, and continue with the next one in line.

In fact this is exactly what `cy.visit` already does under the hood.

We can utilize this same process upon a renderer / browser process crashing to continue on with the run.

So it may look something like this:

```bash
(Running Tests)

✓ test 1 - foo
✓ test 2 - bar
✓ test 3 - baz

Oh noes the renderer process crashed... we will attempt to recover

...Restarting tests at 'test 4 - quux'

✓ test 4 - quux
✓ test 5 - ipsum
```

Taking this a step further, we are starting to see several patterns emerge with how and why renderer processes crash - it is almost always related to extremely long test runs in a memory starved environment (such as Docker).

It may even be a good idea for us to always preemptively "break up" headless runs by spec file. 

In other words, we could have an option like `restartBrowserBetweenSpecFiles: true` which would automatically kill the renderer / browser process before moving on to a different spec file (but still rebuild the state of the UI correctly, and still have a single contiguous video recording). 

To the user it would look like nothing is really different, but internally the renderer process would be killed and then restarted.

This would forcefully purge primed memory from the process, which could keep environments like `docker` from ever crashing to begin with.

Depends on: #6170

**Comment by jheijkoop** [comment]
We actually have these crashes halfway in a single spec and we have stalling too. I tried debugging this with strace and it seems to be constantly trying to acquire some locks. 
Our app seems to make the browser allocate 400+ MB of memory fast and the whole suite can go up to 2 GB...
So resetting between specs might not be enough. Maybe between it/test is also an option?

Setting the `--ipc=host` does fix this, but I wonder what happens if two instances of the test run simultaneously. Could a clash occur?

**Comment by brian-mann** [comment]
How could two instances of the test run occur simultaneously? If you wanted to parallelize you would do it over two different docker containers.

**Comment by jheijkoop** [comment]
yes, two docker instances. It might be a false fear of a clash. I'm completely unaware/ignorant of what the two docker instances do share with `--ipc=host`

**Comment by muslim-niche** [comment]
Hi I am running test cases on aws ec2 small instance and I am having this issue
https://on.cypress.io/renderer-process-crashed
Is there any way to avoid this

**Comment by jheijkoop** [comment]
Did you try the `--ipc=host` fix?

**Comment by muslim-niche** [comment]
But I am not using Docker

**Comment by jheijkoop** [comment]
if not sandboxed, you might have multiple chrome instances fighting over resources. What is you setup? any concurrency? are you open to a different setup?

**Comment by khiettran** [comment]
Any update on this? As Im now getting the error with Chromium usually crashes when running amount of test suites.

**Comment by brian-mann** [comment]
This issue has been superseded by this: https://github.com/cypress-io/cypress/issues/681

That will remove the need to recover since it fixes the problem at its core

**Comment by tizmagik** [comment]
We've started hitting this fairly frequently now too

**Comment by dsherret** [comment]
I'm having this happen randomly on travis-ci with cypress 3.0.2 (I just recently started using cypress so no clue if it happened in a previous version). It might be good to add this flag even with #681 resolved.

**Comment by jheijkoop** [comment]
In hindsight my fix with `--ipc=host` might be related to the shared memory issue I described in https://github.com/cypress-io/cypress/issues/350 and giving the container more shared memory might resolve crashes.

**Comment by jdtzmn** [comment]
I'm also getting this issue now with cypress v3.1.0. Any updates?

**Comment by mechanical-turk** [comment]
Hi cypress team!

We are also getting this error when we use `cypress run` as well as `cypress open`

We noticed that it happens more when we use `cy.wait`. We can consistently reproduce it when we use `cy.wait` with a value greater than `20000`. This is on our circle-ci linux containers fyi.

**Comment by gwaihir8** [comment]
Hi, I'm currently trying to use cypress in Gitlab CI. I Figured out most parts, except the browser craching. 

my current gitlab CI test job is the following:

```
test_dev:
  only:
    - dev
  stage: test
  image: cypress/base:10
  script:
    - npm i --save-dev cypress
    - $(npm bin)/cypress run --reporter junit --reporter-options "mochaFile=results_[hash].xml,toConsole=true"
  artifacts:
    paths:  
      - cypress/videos
    reports:
      junit: results_*.xml
    expire_in: 1 week
```

This works great when the browser doesn't crash, including test reporting in gitlab's merge requests. However, it fails 50% of the times. Using the --ipc=host tag is afaik not an option in Gitlab CI.

**Comment by jheijkoop** [comment]
Have you tried increasing the shared memory instead, like I discribe in https://github.com/cypress-io/cypress/issues/350 ?

**Comment by gwaihir8** [comment]
I am using shared runners on gitlab ci, and shm-size doesn't seem to be an option for shared runners. Thanks anyway

**Comment by jheijkoop** [comment]
I think you can configure it using this documentation https://docs.gitlab.com/runner/configuration/advanced-configuration.html#the-runners-docker-section

**Comment by egucciar** [comment]
Hi Please provide fix / explanation of this issue. It is always happening on one test case (and only one test case). I do not think it has to do with memory but there is no way to know. I was able to reproduce it locally without docker. I think it has to do with origin (subdomain) changes. Thanks

**Comment by ajcann** [comment]
Hi, we're also experiencing this issue in Kubernetes (using Jenkins as our CI engine). Would be happy to provide additional information if helpful.

**Comment by jpike88** [comment]
I'm having this issue often enough to render the framework barely usable. It is likely because our codebase has grown to some sort of critical mass... I'm on CodeBuild and am using the cypress base 10 docker image. Now to figure out how to slot ipc=host in as an option. We've written extensive tests in Cypress, I'm not enjoying being in this position...

**Comment by jpike88** [comment]
After more experimentation, I've found that using the cypress/browsers:chrome69 image instead of the cypress/base:10 made the issue go away. This issue is likely to be tied to an older version of electron being unable to handle a larger codebase, and I think more effort should go into updating electron.

**Comment by mitar** [comment]
One useful thing in meantime would be if Cypress could have some way to communicate this to the caller that the browser failed. Then I could re-run the test inside CI automatically. Maybe an exit code from npm call could be different? Or some other way to determine that tests failed because of Chrome failing and not because of tests failing. Could this be added in meantime? So recovery could then be done outside of Cypress.

**Comment by mitar** [comment]
I think that since this issue has been made there is now a better fix for the problem by asking Chrome not to use `/dev/shm`. I opened #3633 for more details about this.

**Comment by itslenny** [comment]
I'm hitting this issue on a small digital ocean droplet (no docker / container). The test runs perfectly a dozen or so times and then starts crashing with this error. If I reboot the droplet it starts working again then eventually dies. Looks like a memory leak to me.

**Comment by ccorcos** [comment]
There appears to be plenty of memory in my docker container

```
df -h /dev/shm
Filesystem      Size  Used Avail Use% Mounted on
shm              30G  8.0K   30G   1% /dev/shm
```

I'm also unable to figure out how to add the `--ipc=host` flag for my CircleCI build... Doesn't appear to be an option.

**Comment by hhudson** [comment]
> I am using shared runners on gitlab ci, and shm-size doesn't seem to be an option for shared runners. Thanks anyway

I have this same issue.

**Comment by RockChild** [comment]
A few days ago I started facing the same issue regardless no changes were made. It's running on Travis without docker and against a separate app that is not installed in the same code base. 
What interesting, that switching to `--browser chrome` seems to help with it, so looks like it is related to the electron no matter if it is headless or not - in both cases it's failing. However, with chrome, you lose the video recording.
Any progress on this topic? @brian-mann

**Comment by ccorcos** [comment]
I've become very impatient waiting for the Cypress folks to fix these crashing issues. In the meantime, I've created a very similar API using selenium and am having no memory issues. There's no recording of tests, but at least it's reliable. Here's a code snippet for you if you want to try it out.

```ts
import { Builder, ThenableWebDriver, By, WebElement, Key, Condition } from "selenium-webdriver"

/**
 * Wrap any promised coming from the Selenium driver so that we can
 * get stack traces that point to our code.
 */
async function wrapError<T>(p: Promise<T>) {
	const e = new Error()
	e["__wrapError"] = true
	try {
		const result = await p
		// Wait just a little bit in case the browser is about to navigate
		// or something.
		await new Promise(resolve => setTimeout(resolve, 20))
		return result
	} catch (error) {
		if (error["__wrapError"]) {
			throw error
		}
		e.message = error.message
		throw e
	}
}

async function waitFor(
	driver: ThenableWebDriver,
	fn: () => Promise<boolean | object>,
	timeout = 2000
) {
	await driver.wait(
		new Condition("wait", async () => {
			try {
				const result = await fn()
				return Boolean(result)
			} catch (error) {
				return false
			}
		}),
		timeout
	)
}

class Element {
	private promise: Promise<WebElement>
	then: Promise<WebElement>["then"]
	catch: Promise<WebElement>["catch"]

	constructor(
		public driver: ThenableWebDriver,
		promise: Promise<WebElement> | WebElement
	) {
		this.promise = Promise.resolve(promise)
		this.then = this.promise.then.bind(this.promise)
		this.catch = this.promise.catch.bind(this.promise)
	}

	/** Map in the monadic sense. */
	map(fn: (elm: WebElement) => Promise<WebElement | undefined | void>) {
		return new Element(
			this.driver,
			wrapError(
				this.promise.then(async elm => {
					const result = await fn(elm)
					if (result) {
						return result
					} else {
						return elm
					}
				})
			)
		)
	}

	waitFor(fn: (elm: WebElement) => Promise<boolean | object>) {
		return this.map(elm => waitFor(this.driver, () => fn(elm)))
	}

	mapWait(fn: (elm: WebElement) => Promise<WebElement>) {
		return this.waitFor(fn).map(fn)
	}

	click() {
		return this.map(elm => elm.click())
	}

	clear() {
		return this.map(elm => elm.clear())
	}

	type(text: string) {
		return this.map(elm => elm.sendKeys(text))
	}

	enter() {
		return this.map(elm => elm.sendKeys(Key.RETURN))
	}

	backspace() {
		return this.map(elm => elm.sendKeys(Key.BACK_SPACE))
	}

	find(selector: string) {
		return this.mapWait(elm => {
			return elm.findElement(By.css(selector))
		})
	}

	findAll(selector: string) {
		return new Elements(
			this.driver,
			this.promise.then(elm => {
				return waitFor(this.driver, () =>
					elm.findElements(By.css(selector))
				).then(() => {
					return elm.findElements(By.css(selector))
				})
			})
		)
	}

	contains(text: string) {
		return this.mapWait(elm => {
			// TODO: escape text.
			// https://stackoverflow.com/questions/12323403
			return elm.findElement(By.xpath(`//*[contains(text(), '${text}')]`))
		})
	}

	clickText(text: string) {
		return this.contains(text).click()
	}
}

class Elements {
	private promise: Promise<Array<WebElement>>
	then: Promise<Array<WebElement>>["then"]
	catch: Promise<Array<WebElement>>["catch"]

	constructor(
		public driver: ThenableWebDriver,
		promise: Promise<Array<WebElement>> | Array<WebElement>
	) {
		this.promise = Promise.resolve(promise)
		this.then = this.promise.then.bind(this.promise)
		this.catch = this.promise.catch.bind(this.promise)
	}

	/** Map in the monadic sense. */
	map(
		fn: (
			elm: Array<WebElement>
		) => Promise<Array<WebElement> | undefined | void>
	) {
		return new Elements(
			this.driver,
			wrapError(
				this.promise.then(async elms => {
					const result = await fn(elms)
					if (Array.isArray(result)) {
						return result
					} else {
						return elms
					}
				})
			)
		)
	}

	waitFor(fn: (elm: Array<WebElement>) => Promise<boolean | object>) {
		return this.map(elm => waitFor(this.driver, () => fn(elm)))
	}

	mapWait(fn: (elm: Array<WebElement>) => Promise<Array<WebElement>>) {
		return this.waitFor(fn).map(fn)
	}

	clickAll() {
		return this.map(async elms => {
			await Promise.all(elms.map(elm => elm.click()))
		})
	}

	atIndex(index: number) {
		return new Element(
			this.driver,
			wrapError(
				this.promise.then(elms => {
					const elm = elms[index]
					if (!elm) {
						throw new Error("Element not found!")
					}
					return elm
				})
			)
		)
	}
}

export class Browser {
	private promise: Promise<void>
	then: Promise<void>["then"]
	catch: Promise<void>["catch"]

	constructor(public driver: ThenableWebDriver, promise?: Promise<void>) {
		this.promise = Promise.resolve(promise)
		this.then = this.promise.then.bind(this.promise)
		this.catch = this.promise.catch.bind(this.promise)
	}

	visit(route: string) {
		return new Browser(
			this.driver,
			wrapError(
				this.promise.then(async () => {
					await this.driver.get(route)
				})
			)
		)
	}

	rerender() {
		return new Browser(this.driver, wrapError(rerender(this.driver)))
	}

	flushTransactions() {
		return new Browser(this.driver, wrapError(flushTransactions(this.driver)))
	}

	find(selector: string) {
		return new Element(
			this.driver,
			wrapError(
				this.promise
					.then(() => {
						return waitFor(this.driver, async () =>
							this.driver.findElement(By.css(selector))
						)
					})
					.then(() => {
						return this.driver.findElement(By.css(selector))
					})
			)
		)
	}

	getClassName(className: string) {
		return this.find("." + className)
	}

	getTitle() {
		return this.driver.getTitle()
	}

	waitFor(fn: () => Promise<boolean>, timeout = 2000) {
		return new Browser(this.driver, waitFor(this.driver, fn))
	}

	waitToLeave(url: string) {
		return new Browser(
			this.driver,
			wrapError(
				waitFor(
					this.driver,
					async () => {
						const currentUrl = await this.driver.getCurrentUrl()
						return url !== currentUrl
					},
					10000
				)
			)
		)
	}

	waitForRoute(url: string) {
		return new Browser(
			this.driver,
			wrapError(
				waitFor(
					this.driver,
					async () => {
						const currentUrl = await this.driver.getCurrentUrl()
						return url === currentUrl
					},
					10000
				)
			)
		)
	}
}
```

**Comment by nmuth** [comment]
We're seeing this issue crop up on Drone, which also doesn't support the `--ipc=host` option. Our containers already have 16GB memory. Some notes on the behavior:

1. Electron logs an error message when it crashes, but actually fail the test run. Our build is green despite the fact that half the tests caused a renderer crash.

2. Chrome doesn't even log a message—it dies silently and the test run hangs forever.

3. The crash does appear to happen at the exact same time on every run, but it's not clear what we're doing to cause it. Rearranging our test code or skipping certain tests resolves the problem temporarily,  but it always creeps back in.

I haven't contributed to Cypress before, but I'd be willing to take a stab at fixing the problem if someone (@brian-mann ?) can show me where to start. My team has lost a ton of time troubleshooting this and I'd love to put it to bed.

**Comment by jennifer-shehane** [comment]
@nmuth 	Please see our contributing guide on how to start: https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md

Are you using version 3.3.1?

**Comment by nmuth** [comment]
@jennifer-shehane Yup, we're on 3.3.1. I've read the contributing guide. I'm still coming to grips with the code. It looks like the crash handler for Electron is [here](https://github.com/cypress-io/cypress/blob/b8a6baaca38276eaf3d647dfaa6de688676ec612/packages/server/lib/modes/run.co
ffee#L433). Where can I hook in to provide a crash handler for Chrome? Would that be in the `launcher` package?

**Comment by jbinto** [comment]
@RockChild Are you on 3.3.x? I commented in [another thread](https://github.com/cypress-io/cypress/issues/350#issuecomment-499064490) that this seems to have popped up since 3.3.0 dropped about ~2 weeks ago.

**Comment by RockChild** [comment]
@jbinto yeah, looks like it started crashing after upgrade to 3.3.1, so I'll try to downgrade to 3.3.0. Thanks for your insights!

**Comment by bogdan-calapod** [comment]
I switched to `cypress/browsers:chrome69`, changed the package version to `3.3.0` and, with the following build step config in `drone.io`, it seems that the renderer doesn't crash anymore:

```
steps:
  - name: dev-tests
    image: cypress/browsers:chrome69
    shm_size: 4096000000
    mem_limit: 1000000000
    commands:
      - npm ci
      - $(npm bin)/cypress verify
      - $(npm bin)/cypress run

```

**Comment by jbinto** [comment]
@RockChild Downgrading to 3.3.0 (or even 3.2.0) has not resolved this issue for us. 

Similar to you we just started seeing this on or around May 27. No idea what has changed, and we have tried just about everything to fix this. It is gradually getting worse, with almost 100% crash rate today (when it started a few weeks ago it was closer to 5-10%).

Only happening on CircleCI. `/dev/shm` is 30GB there. No pattern to where the tests fail. Nothing interesting when using `DEBUG=cypress:*`.

**Comment by nitzel** [comment]
`If you’re seeing consistent crashes and would like this implemented, please leave a note in the issue.` 
Yes, please.

**Comment by bogdan-calapod** [comment]
Any update on this ? I've tried making a wrapper using Cypress in a .js file but it seems that the renderer errors aren't caught by `Cypress.catch()`

**Comment by ttomaszewski** [comment]
Please fix.

**Comment by neboryte** [comment]
We are hitting this problem as well. Not using Docker.

**Comment by bogdan-calapod** [comment]
I've just given up and switched to testcafe.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by EvanHerman** [comment]
Experiencing consistent browser crashes inside of my Jenkins pipeline. Unable to get around this. I may also have to investigate alternative solutions as this is rather unstable and unpredictable (the reason I moved away from CodeCeption).

Things run great locally, but once we try running the tests on our Jenkins server the browser crashes every time and my tests never pass.

**Comment by jbinto** [comment]
@EvanHerman (and anyone else on this thread): FWIW, since switching to Chrome (from Electron) and setting some flags we have not seen a crash in CI for almost 2 months now.

See https://github.com/cypress-io/cypress/issues/350#issuecomment-503231128 for details.

**Comment by bahmutov** [comment]
And just to add - there is open pull request that adds video recording to Chrome https://github.com/cypress-io/cypress/pull/4791 which is THE main thing stopping people from using Chrome on CI

**Comment by EvanHerman** [comment]
@jbinto Thanks for the tip - I'll get Chrome installed on our VM and test things out. 👍

**Comment by mvandebunt** [comment]
Please fix this. I’m not able to set ipc=host in my ci/cd pipeline

**Comment by darkyndy** [comment]
Please fix this or provide work-around for different environments. In my case I'm running Cypress using Jenkins and pipelines where I do not have access to flags.

**Comment by calwayne** [comment]
Please fix this issue as I hitting 'sad face' error with docker.

**Comment by davidzambrana** [comment]
We're seeing this quite often lately as well with `3.7.0`

**Comment by 9odzilla** [comment]
Happening quite a bit with electron on 3.8

**Comment by mattrlong** [comment]
Update, I've found that since removing the CPU limits from our POD definition this issue is no longer occuring. Whether this is advisable I'm unqualified to say.

**Comment by dialex** [comment]
Migrated from `3.7.0` to `3.8.1` and now we are seeing this error for the first time (when running inside Docker, on GitLab CI)

**Comment by hannahhaken** [comment]
Also seeing this issue across multiple apps which utilise Cypress 3.8, Electron 78 & Docker.

**Comment by hannahhaken** [comment]
We are still seeing this issue for a couple of our apps which utilise Cypress. Others are working fine.

**Comment by sakib-peloton** [comment]
I'm seeing this issue after upgrading cypress. Especially when i run it with percy and take many snapshots.

**Comment by Psycholisk** [comment]
I am facing the same issue as well, I tried optimizing my tests and following the best practices guide:  https://docs.cypress.io/guides/references/best-practices.html. I can see improvements but still crashes the CI most of the time.

**Comment by ameem91** [comment]
I am also facing this issue consistently using headless Electron in CI. I'm on 3.8.1.

**Comment by HaniOsmanRL** [comment]
Also facing this issue. No changes to the code base. It just started to happen on all of our Jenkins job. Please advise!

**Comment by jaroslav-kubicek** [comment]
We started facing this issue today, with no changes, cypress 4.2.0 on Gitlab CI in Electron browser

**Comment by sesamechicken** [comment]
@jaroslav-kubicek I saw that yesterday on Gitlab CI and disabling the video seemed to help. Still digging in.

**Comment by wusgarden** [comment]
We're also seeing the same problem with the same error that @jaroslav-kubicek posted :(

**Comment by sesamechicken** [comment]
We solved our issues by pulling the latest image (`4.3.0`) and making sure to add `  "numTestsKeptInMemory": 1` in our cypress.json file.

**Comment by ttomaszewski** [comment]
We have solved the crash on our end by passing the `--disable-dev-shm-usage` flag to Chrome.
https://docs.cypress.io/guides/guides/continuous-integration.html#In-Docker

**Comment by sesamechicken** [comment]
> We have resolved crashes on our end by passing the `--disable-dev-shm-usage` flag to Chrome in `plugins/index.js`
> https://docs.cypress.io/guides/guides/continuous-integration.html#In-Docker
> 
> ```
> module.exports = (on, config) => {
>   // add --disable-dev-shm-usage chrome flag
>   on('before:browser:launch', (browser, launchOptions) => {
>     if (browser.family === 'chromium') {
>       console.log('Adding Chrome flag: --disable-dev-shm-usage');
>       launchOptions.args.push('--disable-dev-shm-usage');
>     }
>     return launchOptions;
>   });
> };
> ```

We totally did the same thing - completely forgot.

```js
module.exports = (on, config) => {
  on('before:browser:launch', (browser = {}, launchOptions) => {
    if (browser.name === 'chrome') {
      launchOptions.args.push('--disable-dev-shm-usage')
      return launchOptions
    }
  
    return launchOptions
  })
}

```

**Comment by abohannon** [comment]
Also experiencing this issue. Tried both solutions provided by @sesamechicken with no luck. 

Happening during Netlify build w/ the netlify-plugin-cypress plugin. I am just starting to add Cypress to my project and only have a single simple test to run, so the issue shouldn't be many tests exhausting memory.

**Comment by smart625** [comment]
Please fix it, I'm on 4.6

**Comment by JoeSchr** [comment]
also just got it while running cypress github-actions

**Comment by josephzidell** [comment]
Any movement on this?

**Comment by josephzidell** [comment]
@sesamechicken It's happening on `6.3.0`. I plan to upgrade to `6.4.0` next week and will report back then

**Comment by sesamechicken** [comment]
@josephzidell please do. What ci/cd platform are you running this through or is this a local job? I think knowing

- Platform 
- Docker
- Cypress version
- Machine details/OS

Would be a tremendous help to identify the issue.

**Comment by helenou** [comment]
I started running on CI to find out with Cypress v 6.4.0 crash happens on electron displaying the error as "the Chromium Renderer process just crashed". 

Any tip to fix this? I enabled debug with DEBUG=cypress:server:util:process_profiler but not much ideas besides the fact i use a docker image. I thought behaviour might be better looking at other browsers options

**Comment by bahmutov** [comment]
We have noticed often Electron crashing while running tests, but switching to use Chrome browser solved the problem with everything else being equal.

**Comment by josephzidell** [comment]
@sesamechicken Thanks for your patience. A few things came up between then and now. It's still an issue on 6.4.0.

Platform: Docker [`cypress/browsers:node12.16.2-chrome81-ff75`](https://hub.docker.com/layers/cypress/browsers/node12.16.2-chrome81-ff75/images/sha256-65125ec2d789e20a44a9696d46d576cc737a594040853637dad623edd85b50d3?context=explore)
Cypress: 6.4.0
OS: see Docker image above

**Comment by sesamechicken** [comment]
> @sesamechicken Thanks for your patience. A few things came up between then and now. It's still an issue on 6.4.0.
> 
> Platform: Docker [`cypress/browsers:node12.16.2-chrome81-ff75`](https://hub.docker.com/layers/cypress/browsers/node12.16.2-chrome81-ff75/images/sha256-65125ec2d789e20a44a9696d46d576cc737a594040853637dad623edd85b50d3?context=explore)
> Cypress: 6.4.0
> OS: see Docker image above

Awesome. Do you happen to have a repo with an example test to run to reproduce this, or, is it with any test inside of this image?

**Comment by Oumayma-KH** [comment]
After updating to the newest version of Cypress 9.7.0, we started having the crush with running in a Docker container, any updates about a solution for that issur? Thanks!

**Comment by mattvb91** [comment]
Just tried it with 10.0.0 and it cant even get as far as loading the tests before crashing

```
> cypress run --browser chrome

The Test Runner unexpectedly exited via a exit event with signal SIGILL

Please search Cypress documentation for possible solutions:

https://on.cypress.io/

Check if there is a GitHub issue describing this crash:

https://github.com/cypress-io/cypress/issues

Consider opening a new issue.

----------

Platform: linux-x64 (Debian - 10.11)
Cypress Version: 10.0.0
```

**Comment by Oumayma-KH** [comment]
Thank you for your response ! It worked for me !

**Comment by mattvb91** [comment]
> These are the steps i follow to solve this issue:
> 
> 1. I add this code to my cypress plugin folder:
> 
> ```
>     if (browser.name === 'chrome') {
>       launchOptions.args.push('--disable-dev-shm-usage')
>       return launchOptions
>     }
>     return launchOptions
>   }) 
> ```
> 
> 2. i use circle ci for CI/CD. In the jobs section of the cypress test in the config.yml, you add this docker image:
> 
> ```
>    docker:
>        - image: cypress/included:9.4.1
> ```
> 
> * For more info on what the docker image above https://github.com/cypress-io/cypress-docker-images#cypress-docker-images-
> 
> 3. In my packag.json, i updated my `cypres:run` script to `cypress run --browser chrome`
> 4. if your using circle ci, use the recommended cypress orbs

Still no luck even with this. 

Interesting tho is ive put a console log before `    if (browser.name === 'chrome') {` and it doesnt even reach that before crashing. So the process is already killed before it gets as far as even loading the config.

**Comment by estefafdez** [comment]
any update on this issue? we are experiencing it on version 11.2.0. Thanks!

**Comment by LeohsPaixao** [comment]
Browser crashed with me on version 11.2.0, any update ?

**Comment by alesiaviktorchyk** [comment]
Hey @brian-mann. Any update on this issue? We are getting `We detected that the Chromium Renderer process just crashed` almost every time with Cypress 11.2.0. Though everything was ok with Cypress 9.6.0.

**Comment by lakshikawaruni100** [comment]
Experiencing the same issue with 12.1.0. for almost all the tests

**Comment by mi4i0** [comment]
Also started to be experienced chrome crashes in headless mode.
It started to appear after moving from 8.4.0 to Cypress 10.11.0 (moved with one-time migration without migrating to v.9...)
We are running our tests in parallel on machines with 2 cores and 4 Gb RAM
SemaphoreCI as CI service
NodeJS 18.4
Chrome
I found that Cypress somehow had eaten all SWP and RAM memory and crashed right after (second terminal).
![image](https://user-images.githubusercontent.com/18115147/208943952-c2b71bf5-09f1-43c1-9a76-12f68c951c13.png)


Is it any way to check by what Cypress process SWP and RAM were eaten?

**Comment by laerteneto** [comment]
I am also facing the same issue with Cypress 12.3.0 in a Jenkins setup.

**Comment by mattvb91** [comment]
@laerteneto can you try 9.7.0

**Comment by mattvb91** [comment]
Issue is now fixed for me on 12.3.0. I was stuck on 9.7.0 trying to upgrade every version and it never worked. This version for some reason works again. I am using this docker image with it:

```
# FROM cypress/browsers:node17.8.0-chrome99-ff97-slim
# https://github.com/cypress-io/cypress-docker-images/blob/master/browsers/node17.8.0-chrome99-ff97-slim/Dockerfile

FROM cypress/base:18.12.1

USER root

RUN node --version

# Chrome dependencies
RUN apt-get update && \
  apt-get install -y \
  fonts-liberation \
  libcurl4 \
  libcurl3-gnutls \
  libcurl3-nss \
  xdg-utils \
  wget \
  curl \
  # clean up
  && rm -rf /var/lib/apt/lists/* \
  && apt-get clean

# install libappindicator3-1 - not included with Debian 11
RUN wget --no-verbose /usr/src/libappindicator3-1_0.4.92-7_amd64.deb "http://ftp.us.debian.org/debian/pool/main/liba/libappindicator/libappindicator3-1_0.4.92-7_amd64.deb" && \
  dpkg -i /usr/src/libappindicator3-1_0.4.92-7_amd64.deb ; \
  apt-get install -f -y && \
  rm -f /usr/src/libappindicator3-1_0.4.92-7_amd64.deb

# install Chrome browser
RUN wget --no-verbose -O /usr/src/google-chrome-stable_current_amd64.deb "http://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_108.0.5359.94-1_amd64.deb" && \
  dpkg -i /usr/src/google-chrome-stable_current_amd64.deb ; \
  apt-get install -f -y && \
  rm -f /usr/src/google-chrome-stable_current_amd64.deb

# "fake" dbus address to prevent errors
# https://github.com/SeleniumHQ/docker-selenium/issues/87
ENV DBUS_SESSION_BUS_ADDRESS=/dev/null

# a few environment variables to make NPM installs easier
# good colors for most applications
ENV TERM=xterm
# avoid million NPM install messages
ENV npm_config_loglevel=warn
# allow installing when the main user is root
ENV npm_config_unsafe_perm=true

ENV CYPRESS_CACHE_FOLDER=/root/.cache/Cypress

RUN npm i -g npm@9.3.0 cypress@12.3.0 && \
    cypress install && cypress verify && cypress info && \
    npm cache clean --force && \
    apt-get install -y tar && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

# give every user read access to the "/root" folder where the binary is cached
# we really only need to worry about the top folder, fortunately
RUN ls -la /root
RUN chmod 755 /root
RUN chmod -R 755 /root/.cache/Cypress
RUN mkdir -p /root/.config/Cypress/cy/production/proxy
RUN chmod -R 777 /root/.config/Cypress

# versions of local tools
RUN echo  " node version:    $(node -v) \n" \
  "npm version:     $(npm -v) \n" \
  "yarn version:    $(yarn -v) \n" \
  "debian version:  $(cat /etc/debian_version) \n" \
  "Chrome version:  $(google-chrome --version) \n" \
  "Firefox version: $(firefox --version) \n" \
  "Edge version:    $(edge --version) \n" \
  "git version:     $(git --version) \n" \
  "whoami:          $(whoami) \n"
```

**Comment by Neoxrus86** [comment]
> Issue is now fixed for me on 12.3.0. I was stuck on 9.7.0 trying to upgrade every version and it never worked. This version for some reason works again. I am using this docker image with it:
> 
> ```
> # FROM cypress/browsers:node17.8.0-chrome99-ff97-slim
> # https://github.com/cypress-io/cypress-docker-images/blob/master/browsers/node17.8.0-chrome99-ff97-slim/Dockerfile
> 
> FROM cypress/base:18.12.1
> 
> USER root
> 
> RUN node --version
> 
> # Chrome dependencies
> RUN apt-get update && \
>   apt-get install -y \
>   fonts-liberation \
>   libcurl4 \
>   libcurl3-gnutls \
>   libcurl3-nss \
>   xdg-utils \
>   wget \
>   curl \
>   # clean up
>   && rm -rf /var/lib/apt/lists/* \
>   && apt-get clean
> 
> # install libappindicator3-1 - not included with Debian 11
> RUN wget --no-verbose /usr/src/libappindicator3-1_0.4.92-7_amd64.deb "http://ftp.us.debian.org/debian/pool/main/liba/libappindicator/libappindicator3-1_0.4.92-7_amd64.deb" && \
>   dpkg -i /usr/src/libappindicator3-1_0.4.92-7_amd64.deb ; \
>   apt-get install -f -y && \
>   rm -f /usr/src/libappindicator3-1_0.4.92-7_amd64.deb
> 
> # install Chrome browser
> RUN wget --no-verbose -O /usr/src/google-chrome-stable_current_amd64.deb "http://dl.google.com/linux/chrome/deb/pool/main/g/google-chrome-stable/google-chrome-stable_108.0.5359.94-1_amd64.deb" && \
>   dpkg -i /usr/src/google-chrome-stable_current_amd64.deb ; \
>   apt-get install -f -y && \
>   rm -f /usr/src/google-chrome-stable_current_amd64.deb
> 
> # "fake" dbus address to prevent errors
> # https://github.com/SeleniumHQ/docker-selenium/issues/87
> ENV DBUS_SESSION_BUS_ADDRESS=/dev/null
> 
> # a few environment variables to make NPM installs easier
> # good colors for most applications
> ENV TERM=xterm
> # avoid million NPM install messages
> ENV npm_config_loglevel=warn
> # allow installing when the main user is root
> ENV npm_config_unsafe_perm=true
> 
> ENV CYPRESS_CACHE_FOLDER=/root/.cache/Cypress
> 
> RUN npm i -g npm@9.3.0 cypress@12.3.0 && \
>     cypress install && cypress verify && cypress info && \
>     npm cache clean --force && \
>     apt-get install -y tar && \
>     apt-get clean && \
>     rm -rf /var/lib/apt/lists/*
> 
> # give every user read access to the "/root" folder where the binary is cached
> # we really only need to worry about the top folder, fortunately
> RUN ls -la /root
> RUN chmod 755 /root
> RUN chmod -R 755 /root/.cache/Cypress
> RUN mkdir -p /root/.config/Cypress/cy/production/proxy
> RUN chmod -R 777 /root/.config/Cypress
> 
> # versions of local tools
> RUN echo  " node version:    $(node -v) \n" \
>   "npm version:     $(npm -v) \n" \
>   "yarn version:    $(yarn -v) \n" \
>   "debian version:  $(cat /etc/debian_version) \n" \
>   "Chrome version:  $(google-chrome --version) \n" \
>   "Firefox version: $(firefox --version) \n" \
>   "Edge version:    $(edge --version) \n" \
>   "git version:     $(git --version) \n" \
>   "whoami:          $(whoami) \n"
> ```

@mattvb91, Hello! Can you please tell me what commands do you use in package.json to run tests?
I now start up like this:
Dockerfile contains entrypoint.sh file
![2023-01-31_11-49-42](https://user-images.githubusercontent.com/55822441/215702533-d56b8d50-a259-4e19-9141-b32c1ec8557c.png)
entrypoint.sh contains the following code:
![2023-01-31_11-49-31](https://user-images.githubusercontent.com/55822441/215702557-a3a37ac7-1ce0-4839-a1c4-2f2779ba0024.png)
In package.json I use commands like the following:
![2023-01-31_11-56-12](https://user-images.githubusercontent.com/55822441/215702652-c115850c-73fe-410b-9ea1-7e8869b37648.png)
I have little understanding of Docker and programming in general. I will be very grateful for your help.

**Comment by mattvb91** [comment]
@Neoxrus86 I just have package.json:

```
  "scripts": {
    "cy:run": "cypress run --browser chrome"
  }
```

and then in my github actions I mount the folder in with all of my tests and run the npm command above:

```
docker run --add-host="test.cypress:172.17.0.1" --network host --env-file <(env | grep cypress_) -v $(pwd):/app {{MY_CYPRESS_IMAGE_NAME}} /bin/bash -c "cd app/cypress && npm i && npm run cy:run"
``` 

in my case I dont have an entrypoint.sh and run tests on `test.cypress` because I have a multitenant SAAS app that has subdomains so I add a host in there that I can work off.

**Comment by Neoxrus86** [comment]
@mattvb91, Thanks a lot! I'll explore this topic in more detail.

**Comment by yaroslav-shpylyk** [comment]
I'm having this issue with electron right now Cypress `^12.3.0`. Chrome works fine but I'm tied up with electron on project. For some reason it crashes after 2-3 seconds. Nothing from the list helps:
1. `experimentalMemoryManagement: true`
2. `numTestsKeptInMemory: 0/1`
3. 

```
on("before:browser:launch", (browser = {}, launchOptions) => {
    if (browser.name === "electron") {
        launchOptions.args["disable-dev-shm-usage"] = true;
    }

    return launchOptions;
});
```

It fails in simple e2e test:
```
describe("App", () => {
    beforeEach(() => {
        cy.visit("/");
    });

    it("redirects to signup page", () => {
        cy.location("pathname").should("equal", "/signup");
    });
});
```

**Comment by mirobo** [comment]
Is there any progress on @brian-mann 's original suggestion?
See also my comment in the linked issue https://github.com/cypress-io/cypress/issues/3602#issuecomment-1502989253

Ideally Chromium Renderer process crashes would be 
1. reported in mocha (wasn't the case in 10.7.0)
2. retried (as many times as defined in the "runMode.retries" setting)

**Comment by nagash77** [comment]
Hi @yaroslav-shpylyk  can you please open a new issue with your reproducible example?  This issue is quite old so we would need more recent examples to be able to properly diagnose and debug.  I am closing this issue as Won't Fix, but if you are still seeing a similar problem please do open a new issue and we will be happy to investigate.
